<!doctype html>
<html class="no-js" lang="en">
<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>CDS 101: Introduction to Computational and Data Sciences</title>
    <link rel="stylesheet" href="https://archive2019.cds101.com/theme/css/cds101.min.css">
    <link rel="stylesheet" href="https://archive2019.cds101.com/theme/css/custom.css">
      <link rel="stylesheet" href="https://archive2019.cds101.com/theme/css/vendor/katex/katex.min.css">
      <link rel="stylesheet" href="https://archive2019.cds101.com/theme/css/vendor/highlightjs/default.css">
      <link rel="stylesheet" href="https://archive2019.cds101.com/theme/css/vendor/fontawesome/fa-svg-with-js.css">
      <link href="https://archive2019.cds101.com/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="CDS 101: Introduction to Computational and Data Sciences Full Atom Feed" />
</head>  <body>
    <div class="header-grid-container">
<div class="banner-container">
  <header>
    <div class="banner">
      <div class="course-logo">
        <a href="https://archive2019.cds101.com/"><img src="https://archive2019.cds101.com/theme/img/course-logo.svg"></a>
      </div>
      <div class="university-logo">
        <a href="https://gmu.edu"><img src="https://archive2019.cds101.com/theme/img/university-logo.svg"></a>
      </div>
    </div>
  </header>
</div>    </div>
<div class="navbar-container">
  <nav class="navbar">
    <div class="navbar-mobile" data-responsive-toggle="main-menu" data-hide-for="large">
      <button type="button" data-toggle="main-menu"><i class="fas fa-bars"></i><span class="navbar-mobile-title">Menu</span></button>
    </div>

    <div class="navbar-menu" id="main-menu">
      <ul class="navbar-menu-list">
        <li>
          <a class="navbar-home-item" href="https://archive2019.cds101.com/"><i class="fas fa-home" data-fa-transform="grow-8"></i></a>
        </li>
            <li>
              <a class="navbar-menu-item" href="https://archive2019.cds101.com/materials.html">Materials</a>
            </li>
            <li>
              <a class="navbar-menu-item" href="https://archive2019.cds101.com/assignments.html">Assignments</a>
            </li>
            <li>
              <a class="navbar-menu-item" href="https://archive2019.cds101.com/labs.html">Labs</a>
            </li>
          <li class="navbar-social">
            <ul class="navbar-social-list">
                <li>
                  <a class="navbar-social-item" href="https://github.com/jkglasbrenner/cds101-course-materials"><i class="fab fa-github" data-fa-transform="grow-8"></i></a>
                </li>
                <li>
                  <a class="navbar-social-item" href="feeds/all.atom.xml"><i class="fas fa-rss-square" data-fa-transform="grow-8"></i></a>
                </li>
            </ul>
          </li>
      </ul>
    </div>
  </nav>
</div>    <div class="content-grid-container">
    <article class="static-page">
<header class="static-page-header">
  <h1 class="static-page-title">Lab 04 – Tidying your&nbsp;dataset</h1>
</header><div class="static-page-body">
  <h2 id="instructions">Instructions</h2>
<p>Obtain the GitHub repository you will use to complete the lab, which contains a starter file named <span class="monospace">lab04.Rmd</span>. This lab introduces you to the concepts of tidy data and how you can reshape your dataset to take advantage of the <span class="monospace">tidyverse</span> tools. Carefully read the lab instructions and complete the exercises using the provided spaces within your starter file <span class="monospace">lab04.Rmd</span>. Then, when you’re ready to submit, follow the directions in the <a href="#how-to-submit"><i class="fas fa-link"></i>How to submit</a> section&nbsp;below.</p>
<h2 id="tidy-data">Tidy&nbsp;data</h2>
<p>As a review, <a href="https://r4ds.had.co.nz/tidy-data.html"><i class="fas fa-link"></i>the principles of <em>Tidy Data</em></a> are visually represented in the figure&nbsp;below:</p>
<p><img src="https://archive2019.cds101.com/img/tidy_data_schematic.png" title="plot of chunk tidy-data-schematic" alt="plot of chunk tidy-data-schematic" width="80%" style="display: block; margin: auto;" /></p>
<p>The three panels are an illustration of the following three&nbsp;rules,</p>
<ul>
<li><p>Each variable must have its own&nbsp;column.</p></li>
<li><p>Each observation must have its own&nbsp;row.</p></li>
<li><p>Each value must have its own&nbsp;cell.</p></li>
</ul>
<p>It is worth emphasizing that there is a difference between an <em>untidy</em> dataset and a <em>dirty</em> dataset. Tidying a dataset means reshaping it by transposing the rows and columns until the format matches the criteria outlined in the above rules, which then allows us to more easily use the <span class="monospace">ggplot2</span> and <span class="monospace">dplyr</span> functions to analyze and visualize a dataset. Cleaning a dirty dataset means that you are fixing misspellings, data entry errors, and dealing with other irregularities in the raw data. This lab is about showing you how to reshape a dataset to make it&nbsp;tidy.</p>
<h2 id="about-the-dataset">About the&nbsp;dataset</h2>
<p>This week’s dataset is a gene expression dataset that comes from a real-world research project that tested the effect of starvation and growth rate on baker’s yeast<span class="citation" data-cites="Brauer:2007"> [1]</span>. The following quote, sourced from a discussion about this paper<span class="citation" data-cites="Robinson:Cleaning:2015"> [2]</span>, describes the meaning of this dataset&nbsp;well:</p>
<blockquote>
<p>Through the process of gene regulation, a cell can control which genes are transcribed from <span class="caps">DNA</span> to <span class="caps">RNA</span> — what we call being “expressed”. (If a gene is never turned into <span class="caps">RNA</span>, it may as well not be there at all). This provides a sort of “cellular switchboard” that can activate some systems and deactivate others, which can speed up or slow down growth, switch what nutrients are transported into or out of the cell, and respond to other stimuli. A <a href="https://en.wikipedia.org/wiki/DNA_microarray"><i class="fas fa-link"></i>gene expression microarray</a> lets us measure how much of each gene is expressed in a particular condition. We can use this to figure out the function of a specific gene (based on when it turns on and off), or to get an overall picture of the cell’s&nbsp;activity.</p>
<p><a href="http://www.molbiolcell.org/content/19/1/352.abstract"><i class="fas fa-link"></i>Brauer 2008</a> used microarrays to test the effect of starvation and growth rate on baker’s yeast <a href="https://en.wikipedia.org/wiki/Saccharomyces_cerevisiae"><i class="fas fa-link"></i>S. cerevisiae</a>, a popular model organism for studying molecular genomics because of its simplicity). Basically, if you give yeast plenty of nutrients (a rich media), except that you sharply restrict its supply of one nutrient, you can control the growth rate to whatever level you desire (we do this with a tool called a <a href="https://en.wikipedia.org/wiki/Chemostat"><i class="fas fa-link"></i>chemostat</a>). For example, you could limit the yeast’s supply of glucose (sugar, which the cell metabolizes to get energy and carbon), of leucine (an essential amino acid), or of ammonium (a source of&nbsp;nitrogen).</p>
<p>“Starving” the yeast of these nutrients lets us find genes&nbsp;that:</p>
<ul>
<li><strong>Raise or lower their activity in response to growth rate</strong>. Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to&nbsp;stress.</li>
<li><strong>Respond differently when different nutrients are being limited</strong>. These genes may be involved in the transport or metabolism of those&nbsp;nutrients.</li>
</ul>
</blockquote>
<p>The table below provides descriptions about the dataset’s 40&nbsp;variables,</p>
<table>
<colgroup>
<col style="width: 17%" />
<col style="width: 82%" />
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="monospace"><span class="caps">GID</span></span></td>
<td>One kind of <span class="caps">ID</span> for each gene</td>
</tr>
<tr class="even">
<td><span class="monospace"><span class="caps">YORF</span></span></td>
<td>Yeast Open Reading Frame (a kind of sub-unit of genetic information)</td>
</tr>
<tr class="odd">
<td><span class="monospace"><span class="caps">NAME</span></span></td>
<td>See below</td>
</tr>
<tr class="even">
<td><span class="monospace"><span class="caps">GWEIGHT</span></span></td>
<td>The paper doesn’t make this clear, but all entries are 1</td>
</tr>
<tr class="odd">
<td><span class="monospace">[<span class="caps">GNP</span>][0.05 ≤ x ≤ 0.30]</span></td>
<td>The letters <span class="monospace">G</span>, <span class="monospace">N</span>, and <span class="monospace">P</span> represent the restricted nutrient. The decimal value is the yeast growth rate.</td>
</tr>
</tbody>
</table>
<p>Note that the <span class="monospace"><span class="caps">NAME</span></span> column contains the following information separated by the double bar <span class="monospace">||</span>&nbsp;symbols:</p>
<table>
<colgroup>
<col style="width: 25%" />
<col style="width: 74%" />
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Gene name</td>
<td>for example, <span class="caps">SFB2</span>. Not all genes have a name here.</td>
</tr>
<tr class="even">
<td>Biological process</td>
<td>for example, “proteolysis and peptidolysis”</td>
</tr>
<tr class="odd">
<td>Molecular function</td>
<td>for example, “metalloendopeptidase activity”</td>
</tr>
<tr class="even">
<td>Systematic <span class="caps">ID</span></td>
<td>for example, <span class="caps">YNL049C</span>. Every gene has one of these unique IDs.</td>
</tr>
<tr class="odd">
<td>Unknown <span class="caps">ID</span> Number</td>
<td>for example, 1082129. The paper doesn’t explain what these mean.</td>
</tr>
</tbody>
</table>
<div class="callout primary">
<p><span class="h4">Note</span></p>
<p>Like in previous labs, it’s recommended that you take a first look at the dataset by viewing it by running <code>View(brauer)</code> in your <em>Console</em>&nbsp;window.</p>
</div>
<h2 id="the-tidyr-package">The <span class="monospace">tidyr</span>&nbsp;package</h2>
<p>Reshaping the gene expression dataset will require us to use two functions found in the <span class="monospace">tidyr</span> package, <code>gather()</code> and <code>separate()</code>. Let’s review how each of these functions works with the extended example from Chapter 12.6 in the <a href="http://r4ds.had.co.nz/"><i class="fas fa-link"></i><em>R for Data Science</em></a>&nbsp;textbook.</p>
<p>Running the <code>library(tidyverse)</code> command at the top of our RMarkdown file loads many packages and example datasets for us, which includes a dataset from the World Health Organization that is stored in the variable <code>who</code>. The first few lines of the <code>who</code> dataset&nbsp;are:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
country
</th>
<th style="text-align:left;">
iso2
</th>
<th style="text-align:left;">
iso3
</th>
<th style="text-align:left;">
year
</th>
<th style="text-align:left;">
new_sp_m014
</th>
<th style="text-align:left;">
new_sp_m1524
</th>
<th style="text-align:left;">
new_sp_m2534
</th>
<th style="text-align:left;">
new_sp_m3544
</th>
<th style="text-align:left;">
new_sp_m4554
</th>
<th style="text-align:left;">
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1980
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1981
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1982
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1983
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1984
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1985
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
<span class="caps">NA</span>
</td>
<td style="text-align:left;">
…
</td>
</tr>
<tr>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
</tr>
</tbody>
</table>
<h3 id="using-gather-to-transpose-columns-into-rows">Using <span class="monospace">gather()</span> to transpose columns into&nbsp;rows</h3>
<p>In this data frame, the names of the 57 columns starting with <span class="monospace">new_sp_m014</span> and ending with <span class="monospace">newrel_f65</span> each refer to a set of three categories, violating the first rule for tidy data. This can easily be fixed by transposing these columns into rows using the <code>gather()</code> command:</p>
<pre class="r"><code>who1 &lt;- who %&gt;% 
  gather(
    new_sp_m014:newrel_f65,
    key = &quot;key&quot;,
    value = &quot;cases&quot;,
    na.rm = TRUE
  )</code></pre>
<p>After applying the <code>gather()</code> operation, the first few rows in the dataset now look as&nbsp;follows:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
country
</th>
<th style="text-align:left;">
iso2
</th>
<th style="text-align:left;">
iso3
</th>
<th style="text-align:left;">
year
</th>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
cases
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1997
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1998
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1999
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2000
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
52
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2001
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
129
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2002
</td>
<td style="text-align:left;">
new_sp_m014
</td>
<td style="text-align:left;">
90
</td>
</tr>
<tr>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
</tr>
</tbody>
</table>
<p>As you can see, we’ve taken the 57 category columnms and converted them into categories underneath a single column named <span class="monospace">key</span> with their corresponding values placed underneath the column <span class="monospace">cases</span>.</p>
<div class="callout primary">
<p><span class="h4">Hint</span></p>
<p>To summarize, the syntax for <code>gather()</code> is as&nbsp;follows:</p>
<pre class="r"><code>dataset %&gt;%
  gather(
    ...,           #  Columns you want to gather into rows
    key = &quot;...&quot;,   #  Variable for storing names of gathered columns
    value = &quot;...&quot;  #  Variable for values stored under gathered columns
  )</code></pre>
</div>
<h3 id="using-separate-to-split-one-column-into-many">Using <span class="monospace">separate()</span> to split one column into&nbsp;many</h3>
<p>The other function you will need to use is <code>separate()</code>. This function takes values in a single column and splits them out into multiple columns. This is used when you have a dataset that doesn’t follow the third rule of tidy&nbsp;data.</p>
<p>Going back to the <span class="monospace">who</span> example, after you gathered the 57 columns into a single column, we might wonder what the values under <span class="monospace">key</span> mean. If you run <code>?who</code>, you’ll see that the underscores <span class="monospace">…</span> separate different variable values. This means that the <span class="monospace">key</span> column contains 3 values per cell, not one, so this is a case for using <code>separate()</code>.</p>
<p>To separate the columns, we run the&nbsp;following:</p>
<pre class="r"><code>who2 &lt;- who1 %&gt;%
  mutate(
    key = str_replace(key, &quot;newrel&quot;, &quot;new_rel&quot;)
  ) %&gt;%
  separate(
    col = key,
    into = combine(&quot;new&quot;, &quot;type&quot;, &quot;sexage&quot;),
    sep = &quot;_&quot;
  )</code></pre>
<p>You can ignore the second line with a <code>mutate()</code> command for now, as this just fixes the missing underscore that you need to do before running&nbsp;separate.</p>
<p>The first few lines of the separated dataset&nbsp;are:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
country
</th>
<th style="text-align:left;">
iso2
</th>
<th style="text-align:left;">
iso3
</th>
<th style="text-align:left;">
year
</th>
<th style="text-align:left;">
new
</th>
<th style="text-align:left;">
type
</th>
<th style="text-align:left;">
sexage
</th>
<th style="text-align:left;">
cases
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1997
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1998
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
1999
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2000
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
52
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2001
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
129
</td>
</tr>
<tr>
<td style="text-align:left;">
Afghanistan
</td>
<td style="text-align:left;">
<span class="caps">AF</span>
</td>
<td style="text-align:left;">
<span class="caps">AFG</span>
</td>
<td style="text-align:left;">
2002
</td>
<td style="text-align:left;">
new
</td>
<td style="text-align:left;">
sp
</td>
<td style="text-align:left;">
m014
</td>
<td style="text-align:left;">
90
</td>
</tr>
<tr>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
<td style="text-align:left;">
…
</td>
</tr>
</tbody>
</table>
<p>As you can see, this has successfully split our one column into&nbsp;three.</p>
<div class="callout primary">
<p><span class="h4">Note</span></p>
<p>To summarize, the syntax for <code>separate()</code> is as&nbsp;follows:</p>
<pre class="r"><code>dataset %&gt;%
  separate(
    col = ...,            #  Name of column to separate
    into = combine(...),  #  Names for new columns formed after separation
    sep = ...,            #  Specifies the separator symbols
    convert = ...         #  If TRUE, tries to set data type for new columns
  )</code></pre>
<p>Your should note that the <span class="monospace">into</span> keyword needs to be specified as a vector of strings that you can create using <code>combine()</code>.</p>
</div>
<h2 id="prune-the-dataset">Prune the&nbsp;dataset</h2>
<p>When dealing with larger datasets that have many columns, it is helpful if you can identify any columns that are extraneous and will not be used for analysis. This shrinks the overall size of the dataset so that it takes up less space in the computer’s memory. For this dataset, the <span class="monospace"><span class="caps">GID</span></span>, <span class="monospace"><span class="caps">YORF</span></span>, and <span class="monospace"><span class="caps">GWEIGHT</span></span> columns aren’t particularly important for any kind of analysis, so they can be&nbsp;removed.</p>
<ol class="example" type="1">
<li>Remove the <span class="monospace"><span class="caps">GID</span></span>, <span class="monospace"><span class="caps">YORF</span></span>, and <span class="monospace"><span class="caps">GWEIGHT</span></span> variables from the dataset and assign the result to the variable <span class="monospace">brauer2</span>.</li>
</ol>
<h2 id="enforcing-one-observation-per-row">Enforcing one observation per&nbsp;row</h2>
<p>Now we will employ a systematic approach while tidying this dataset by going through the rules one by one and determining what (if anything) needs to be reshaped. We start with the first rule that “each variable must have its own column.” The columns <span class="monospace">G0.05</span> through <span class="monospace">U0.3</span> each contain the results of different experimental trials, telling us which nutrient was restricted and what growth rate was selected for the yeast. Knowing this, we conclude that this dataset violates the first rule, as each row contains multiple experimental&nbsp;trials.</p>
<ol start="2" class="example" type="1">
<li>Use <code>gather()</code> to transpose the columns <span class="monospace">G0.05</span> through <span class="monospace">U0.3</span> into a series of rows to satisfy the rule that &#8220;each variable must have its own column. Use the label <span class="monospace">sample</span> for the new categorical variable created by <code>gather()</code> that contains the gathered column names and use the label <span class="monospace">expression</span> for the new numerical variable containing the values originally under the <span class="monospace">G0.05</span> through <span class="monospace">U0.3</span> columns. Assign your result to the variable <span class="monospace">brauer3</span>.</li>
</ol>
<p>After completing this step, the <em>one observation per row</em> rule will be&nbsp;satisfied.</p>
<h2 id="enforcing-one-column-per-variable-and-one-value-per-cell">Enforcing one column per variable and one value per&nbsp;cell</h2>
<p>The second and third rules are intertwined, and so fixing one will help to satisfy the other. Looking at the version of the dataset in <span class="monospace">brauer3</span>, we can clearly see that the <em>one value per cell</em> rule is violated by the <span class="monospace"><span class="caps">NAME</span></span> and <span class="monospace">sample</span> columns. These are violations that can be fixed using the <code>separate()</code> function. Let’s separate the <span class="monospace">sample</span> column we created with the <code>gather()</code> function&nbsp;first.</p>
<div class="callout primary">
<p><span class="h4">Hint</span></p>
<p>If you don’t want the <code>separate()</code> function to delete any individual letters or symbols when splitting a column, you can set the <span class="monospace">sep</span> argument equal to an integer, for example <span class="monospace">sep = 1</span>. This tells <code>separate()</code> to split the column right after the first&nbsp;letter.</p>
</div>
<ol start="3" class="example" type="1">
<li>Separate the <span class="monospace">sample</span> column that you created in <strong>Exercise 2</strong> into two variables, <span class="monospace">nutrient</span> and <span class="monospace">rate</span>. After separation, the data type for the <span class="monospace">rate</span> column should be <em>double</em> (<span class="monospace">&lt;dbl&gt;</span>), not <em>character</em>. Assign your result to the variable <span class="monospace">brauer4</span>.</li>
</ol>
<p>Now we turn to the <span class="monospace"><span class="caps">NAME</span></span> column where each value is separated by two bars <span class="monospace">||</span> and there are five variables contained in each cell. Using the information from the <a href="#about-the-dataset"><i class="fas fa-link"></i>about the dataset</a> section, we deduce that we should name the new columns <span class="monospace">gene_name</span>, <span class="monospace">biological_process</span>, <span class="monospace">molecular_function</span>, <span class="monospace">systematic_id</span>, and <span class="monospace">number</span>.</p>
<div class="callout secondary">
<p><span class="h4">Important!</span></p>
<p>If you try to use <span class="monospace">sep = “||”</span> in <code>separate()</code>, it will not work as expected. By default, the bar symbol <span class="monospace">|</span> is interpreted as the boolean operator <span class="caps">OR</span>, which should be familiar from creating rules for <code>filter()</code>. To get around this, we can preceed <span class="monospace">|</span> with backslashes to signal to R that we want it to read <span class="monospace">|</span> as a symbol and not as the boolean operator <span class="caps">OR</span>. Thus, we should use <span class="monospace">sep = “&#92;&#92;|&#92;&#92;|”</span> to define our&nbsp;separator.</p>
</div>
<ol start="4" class="example" type="1">
<li>Use <code>separate()</code> to split the values in the <span class="monospace"><span class="caps">NAME</span></span> column into five columns named <span class="monospace">gene_name</span>, <span class="monospace">biological_process</span>, <span class="monospace">molecular_function</span>, <span class="monospace">systematic_id</span>, and <span class="monospace">number</span>. Assign your result to the variable <span class="monospace">brauer5</span>.</li>
</ol>
<p>We’re nearly done! If you inspect the dataset with <code>View(brauer5)</code>, you’ll notice that the text under some of the columns has quotation marks with unnecessary blank space. For example, <span class="monospace">&#8221; molecular function unknown &#8220;</span>. Removing this blank space (typically called “white space”) is called <strong>trimming</strong>, and we can use the <code>str_trim()</code> function for this&nbsp;purpose.</p>
<ol start="5" class="example" type="1">
<li><p>Remove the whitespace from the five new columns you created using <code>mutate_at()</code> in combination with <code>str_trim()</code>:</p>
<pre class="r"><code>brauer5 %&gt;%
  mutate_at(
    vars(gene_name:systematic_id),
    str_trim
  )</code></pre>
<p>Assign your result to the variable <span class="monospace">brauer_tidy</span>.</p></li>
</ol>
<p>Now we’re&nbsp;done!</p>
<h2 id="visualizations-using-the-tidy-dataset">Visualizations using the tidy&nbsp;dataset</h2>
<p>Having a tidy dataset makes it easy to use the rest of the <span class="monospace">tidyverse</span> packages, such as <span class="monospace">ggplot2</span>. To demonstrate this, we will now create a plot that a biologist would use to explore gene expression trends that emerge due to restricting different nutrients and enforcing different growth&nbsp;rates.</p>
<ol start="6" class="example" type="1">
<li>Filter <span class="monospace">brauer_tidy</span> so that it only contains entries with <span class="monospace">gene_name</span> equal to <span class="monospace"><span class="caps">LEU1</span></span>. Then, create a line plot where you place <span class="monospace">rate</span> on the horizontal axis, <span class="monospace">expression</span> on the vertical axis, and assign different colors to <span class="monospace">nutrient</span>.</li>
</ol>
<h2 id="additional-exercises">Additional&nbsp;exercises</h2>
<ol start="7" class="example" type="1">
<li><p>What are the dataset’s dimensions (the number of rows and columns) after completing the tidying procedure? Do you find it easier to read the data in this format, or did you think the original format was easier to understand? Why or why&nbsp;not?</p></li>
<li><p>In <strong>Exercise 6</strong> we were able to easily create a line plot of <span class="monospace">expression</span> as a function of <span class="monospace">rate</span> for the different kinds of nutrients for one of the genes in the microarray. What specifically did the tidying procedure do to make it simple for us to create this plot? Could we still create the same plot with <span class="monospace">ggplot2</span> using the untidy version of the dataset? If so, try to write the <span class="monospace">ggplot2</span> code for it. If not, identify what it is about the untidy dataset that prevents you from creating the same&nbsp;plot.</p></li>
</ol>
<h2 id="how-to-submit">How to&nbsp;submit</h2>
<p>To submit your lab, follow the two steps below. Your lab will be graded for credit <strong>after</strong> you’ve completed both&nbsp;steps!</p>
<ol type="1">
<li><p>Save, commit, and push your completed R Markdown file so that everything is synchronized to GitHub. If you do this right, then you will be able to view your completed file on the GitHub&nbsp;website.</p></li>
<li><p>Knit your R Markdown document to the <span class="caps">PDF</span> format, export (download) the <span class="caps">PDF</span> file from RStudio Server, and then upload it to <em>Lab 4</em> posting on&nbsp;Blackboard.</p></li>
</ol>
<h2 id="cheatsheets">Cheatsheets</h2>
<p>You are encouraged to review and keep the following cheatsheets handy while working on this&nbsp;lab:</p>
<ul>
<li><p><a href="https://github.com/rstudio/cheatsheets/raw/master/data-import.pdf">data import/tidyr&nbsp;cheatsheet</a></p></li>
<li><p><a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf">dplyr&nbsp;cheatsheet</a></p></li>
<li><p><a href="https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf">ggplot2&nbsp;cheatsheet</a></p></li>
<li><p><a href="https://github.com/rstudio/cheatsheets/raw/master/rstudio-ide.pdf">RStudio&nbsp;cheatsheet</a></p></li>
<li><p><a href="https://github.com/rstudio/cheatsheets/raw/master/rmarkdown-2.0.pdf">R Markdown&nbsp;cheatsheet</a></p></li>
<li><p><a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">R Markdown&nbsp;reference</a></p></li>
</ul>
<h2 id="credits">Credits</h2>
<p>This lab is licensed under a <a href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>. Exercises and instructions written by James Glasbrenner for <span class="caps">CDS</span>-102.</p>
<h2 id="references" class="unnumbered">References</h2>
<div id="refs" class="references">
<div id="ref-Brauer:2007">
<p>[1] <span class="caps">M. J.</span> Brauer, C. Huttenhower, <span class="caps">E. M.</span> Airoldi, R. Rosenstein, <span class="caps">J. C.</span> Matese, D. Gresham, <span class="caps">V. M.</span> Boer, <span class="caps">O. G.</span> Troyanskaya, and D. Botstein, “Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast,” Molecular Biology of the Cell <strong>19</strong>, 352&nbsp;(2007).</p>
</div>
<div id="ref-Robinson:Cleaning:2015">
<p>[2] David Robinson, “Cleaning and Visualizing Genomic Data: A Case Study in Tidy Analysis,” Variance Explained&nbsp;(2015).</p>
</div>
</div>

</div>    </article>
    </div>
<footer class="page-footer">
    <div class="footer-grid-container">
      Except where otherwise noted, site materials are created by
      <a href="https://cos.gmu.edu/cds/faculty-profile-james-glasbrenner/">
        James K. Glasbrenner
      </a>
      and licensed under a
      <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">
        Creative Commons Attribution-ShareAlike 4.0 International License
      </a>.
    </div>
</footer><script src="https://archive2019.cds101.com/theme/js/cds101.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/highlightjs/highlight.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/katex/katex.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/katex/contrib/auto-render.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/wookmark/wookmark.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/fontawesome/fontawesome-all.min.js"></script>
  <script src="https://archive2019.cds101.com/theme/js/vendor/vendorInit.js"></script>
  </body>
</html>